crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1166,18 +1179,140 @@ impl<'db> TypeInferenceBuilder<'db> {
+        kind: AssignmentKind,
         definition: Definition<'db>,
     ) {
-        let expression = self.index.expression(assignment.value.as_ref());
+        let expression = self.index.expression(value);
         let result = infer_expression_types(self.db, expression);
         self.extend(result);
-        let value_ty = self.expression_ty(&assignment.value);
-        self.add_binding(assignment.into(), definition, value_ty);
+
+        let value_ty = self.expression_ty(value);
+
+        let target_ty = match (value_ty, kind) {
+            (
+                Type::Tuple(_) | Type::StringLiteral(_) | Type::LiteralString,
+                AssignmentKind::Sequence,
+            ) => self.infer_sequence_unpacking(target, value_ty, variable),
+            (_, AssignmentKind::Sequence) => value_ty
+                .iterate(self.db)
+                .unwrap_with_diagnostic(value.into(), self),
+            _ => value_ty,
+        };
+
+        self.add_binding(variable.into(), definition, target_ty);
         self.types
             .expressions
-            .insert(target.scoped_ast_id(self.db, self.scope), value_ty);
+            .insert(variable.scoped_ast_id(self.db, self.scope), target_ty);
+    }
+
+    fn infer_sequence_unpacking(
+        &mut self,
+        target: &ast::Expr,
+        value_ty: Type<'db>,
+        variable: &ast::ExprName,
+    ) -> Type<'db> {
+        // The inner function is recursive and only differs in the return type which is an `Option`
+        // where if the variable is found, the corresponding type is returned otherwise `None`.
+        fn inner<'db>(
+            builder: &mut TypeInferenceBuilder<'db>,
+            target: &ast::Expr,
+            value_ty: Type<'db>,
+            variable: &ast::ExprName,
+        ) -> Option<Type<'db>> {
+            match target {
+                ast::Expr::Name(name) if name == variable => {
+                    return Some(value_ty);
+                }
+                ast::Expr::Starred(ast::ExprStarred { value, .. }) => {
+                    // TODO: Wrap the `value_ty` in a list type.
+                    return inner(builder, value, Type::Todo, variable);
+                }
+                ast::Expr::List(ast::ExprList { elts, .. })
+                | ast::Expr::Tuple(ast::ExprTuple { elts, .. }) => match value_ty {
+                    Type::Tuple(tuple_ty) => {
+                        let starred_index = elts.iter().position(ast::Expr::is_starred_expr);
+
+                        let element_types = if let Some(starred_index) = starred_index {
+                            if tuple_ty.len(builder.db) >= elts.len() - 1 {
+                                let mut element_types = Vec::with_capacity(elts.len());
+                                element_types.extend_from_slice(
+                                    // SAFETY: Safe because of the length check above.
+                                    tuple_ty.elements(builder.db).get(..starred_index).unwrap(),
```

Comment:
```suggestion
                                    &tuple_ty.elements(builder.db)[..starred_index],
```

Possibly, depending on what `elements` returns